dockershim: Return Labels as Info in ImageStatus.

shlevy opened the pull request:
c6ddc749e85f4281628f366d65cf924c22d6cd99 added an Info field to
ImageStatusResponse when Verbose is true. This makes the image's
Labels available in that field, rather than unconditionally returning
an empty map.

**What this PR does / why we need it**:

This PR exposes an image's `Labels` through the CRI. In particular, I want this so I can write an `ImageService` wrapper that delegates all operations to a real `ImageService` but also, when the right `Labels`, ensures any needed [nix store](https://nixos.org/nix/) paths are present on the system when an image is pulled, enabling users to use nix for package distribution while still using containers for isolation and kubernetes for orchestration. In general, though, this should be useful for anything that wants to know about an image's `Labels`

**Special notes for your reviewer**:

I'd prefer to put this change into the `Image` protobuf type instead of putting it into `Info` (gated by `Verbose` or not, available in other requests like `ListImages` or not), but that would be a change to the protocol and it seems `Info` was introduced exactly for this purpose. If it's acceptable to put this into `Image`, I'll rework this.

If this change is acceptable, I will also do the work for `cri-o`, `rktlet`, `frakti`, and `cri-containerd` where applicable.

I have started the process for my employer to sign on to the CLA. I don't have reason to expect it to take long, but because there is more work to do if this change is desired I'd prefer if we can start review before that is completed.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
dockershim now makes an Image's Labels available in the Info field of ImageStatusResponse
```

cblecker commented:
/ok-to-test

cblecker commented:
@shlevy Please review and sign the CLA: https://github.com/kubernetes/kubernetes/pull/58036#issuecomment-356478293

shlevy commented:
@cblecker I am waiting on my employer's guidelines for signing as an individual or as a corporation

shlevy commented:
OK, the process is started by my employer and the CLA should be signed within a week or so. In the mean time, though, if this work is desirable there's more work I'd need to do for the other CRI server implementations, so if possible a high level review (taking into account my "notes for reviewer") would be appreciated. I understand if everything needs to wait for the CLA though!

mtaufen commented:
/unassign
/assign @yujuhong

shlevy commented:
OK, CLA finally came through. Rebasing now.

shlevy commented:
/retest

Commit:
dockershim: Return Labels as Info in ImageStatus.

c6ddc749e85f4281628f366d65cf924c22d6cd99 added an Info field to
ImageStatusResponse when Verbose is true. This makes the image's
Labels available in that field, rather than unconditionally returning
an empty map.

shlevy commented:
@cblecker How do I get the CLA check to run again?

cblecker commented:
@shlevy The CLA check runs whenever the PR is updated. It's still reporting that you don't have a signed CLA. You may wish to check the top comment one more time to confirm you've followed all the needed steps.

shlevy commented:
OK, I *think* CLA is finally all covered :sweat_smile:

shlevy commented:
Hmm, I don't really understand what's going on in that test failure, is that really related to this PR?

cblecker commented:
/test pull-kubernetes-unit, flake #59426

cblecker commented:
/retest

yujuhong commented:
I'm okay with the change since it's gated by "verbose".
Each runtime implements the verbose response differently, so they can determine whether this is useful for them.

/lgtm

shlevy commented:
@yujuhong Any thoughts on this?

> I'd prefer to put this change into the `Image` protobuf type instead of putting it into `Info` (gated by `Verbose` or not, available in other requests like `ListImages` or not), but that would be a change to the protocol and it seems `Info` was introduced exactly for this purpose. If it's acceptable to put this into `Image`, I'll rework this.

Random-Liu commented:
@yujuhong This seems fine to me. :)

@shlevy I think we can consider adding label as CRI fields when we have more use cases. The current use case seems more like an implementation detail for a specific use case. :)

ironhouzi commented:
@Random-Liu This feature is sorely needed by my use case:

We use containers for blackboxing trusted algorithms or data processors basically. We store definitions of inputs, how they map to commandline arguments to the executable and what outputs the processor is expected to produce. It makes a lot of sense to have this metadata as image labels instead of files inside the image, since the metadata is only needed by the container scheduler and not by the container runtime. Having the metadata and image bundled together is very nice in terms of portability.

shlevy commented:
@Random-Liu What would be the appropriate process for proposing adding this as an official field of the `Image` protobuf type?
